bluesky/run_engine.py

```diff
@@ -91,6 +91,10 @@ def __set__(self, obj, value):
      Parameter('subs', Parameter.POSITIONAL_ONLY, default=None),
      Parameter('metadata_kw', Parameter.VAR_KEYWORD)])
 
+#Default scan_id_source
+def default_scan_id_source(md):
+    scan_id = md.get('scan_id', 0) + 1
+    md['scan_id'] = scan_id
```

Comment:
I'd like to suggest a subtle but important change.

As written here, `default_scan_id_source` is a function that _mutates its parameter_. Sometimes this is a useful thing to be able to do. For example, you can find places where we pass a cache of previously-computed values into a function, saying, "Use this cache to speed up your computation, and feel free to add things to it (mutate it)." In that case, mutating the parameter is a simple way to get the behavior we need.

But other that special tricks like that, it's best to avoid mutating parameters. Functions should generally take in objects and return other objects, without mutating the objects they took it. Mutation can lead to confusing behavior if the caller does not expect the function to go and change the parameters it was given. For example, suppose I were to write tests for this function and, in doing so, call the function multiple times. I would have take care to make a fresh copy of the input `md` each time because calling the function modifies my input. To avoid this, you can make a version of this function that _returns_ the new `scan_id` rather than updating the `md` dictionary in place. Then leave it up to the caller (i.e. the RunEngine) to do the mutation, as in `md['scan_id'] = self.scan_id_source(self.md)`.

Let me know if I should give more examples and I will try to think of some!

Reply:
Please check if I fully understand by code review
After edited code, I think it do make sense for people to understand  `md['scan_id'] = self.scan_id_source(self.md)`